skbio/tree/_tree.py

```diff
@@ -3053,67 +3069,500 @@ def count(self, tips=False):
+            branch lengths will be replaced with 0. Default is False.
+
+            .. versionadded:: 0.6.3
 
         Returns
         -------
-        frozenset
-            A frozenset of frozensets of str
+        frozenset of frozenset of str, or
+            All subsets of taxa defined by nodes descending from self. Returned if
+            `map_to_length` is False.
+        dict of {frozenset of str: float}
+            Mapping of all subsets of taxa to their branch lengths. Returned if
+            `map_to_length` is True.
 
         See Also
         --------
         subset
         compare_subsets
+        biparts
+
+        Notes
+        -----
+        The returned value represents the tree as a set of nested sets, each of which
+        representing a clade in the tree. It is useful for assessing topological
+        patterns of a tree.
+
+        The returned value itself and each of its components (frozensets) are unordered
+        and hashable, making it efficient for lookup and comparison. For example, one
+        can check whether a group of taxa form a clade in the tree, regardless of its
+        internal structure.
+
+        This method can be applied to both rooted and unrooted trees. However, the
+        underlying assumption is that the direction of descendance is from the current
+        node to the tips below. That is, the root of the tree, even if not explicitly
+        defined, should be at or above the current node. This should be considered when
+        applying this method to an unrooted tree. If such an assumption is not present,
+        one should consider using :meth:`biparts` instead.
+
+        This method operates on the subtree below the current node.
 
         Examples
         --------
         >>> from skbio import TreeNode
-        >>> tree = TreeNode.read(["(((a,b)c,(d,e)f)h)root;"])
+        >>> tree = TreeNode.read(["((a,(b,c)d)e,(f,g)h)i;"])
+        >>> print(tree.ascii_art())
+                            /-a
+                  /e-------|
+                 |         |          /-b
+                 |          \d-------|
+        -i-------|                    \-c
+                 |
+                 |          /-f
+                  \h-------|
+                            \-g
+
         >>> subsets = tree.subsets()
-        >>> len(subsets)
-        3
+        >>> for s in sorted(subsets, key=sorted):
+        ...     print(sorted(s))
+        ['a', 'b', 'c']
+        ['b', 'c']
+        ['f', 'g']
+
+        >>> {'a', 'b', 'c'} in subsets
+        True
+
+        >>> {'a', 'b'} in subsets
+        False
+
+        """
+        if within and not isinstance(within, (set, frozenset, dict)):
+            within = frozenset(within)
+
+        # initiate result
+        if map_to_length:
+            subsets = {}
+            subsets_get = subsets.get
+        else:
+            subsets = []
+            subsets_append = subsets.append
+
```

Comment:
I see the rationale here and your comment above (`if node.name in within:`). However, I think that it will be more performant to skip checking when `within` is None. Even hash table search is _O_(1), it is still not zero, especially when the tree is huge.

The most performant approach would be separating `within is None` and `within is not None` into two independent loops. However, this can make the code lengthy and hard to maintain, especially when this method also has other flags (i.e., multiple combinations of flags).

While I was coding, I benchmarked this scenario and others, and found that have an `if within` check in the loop has minimum impact on the performance.

Reply:
Tested and it seems that the class idea is slower.

```python
>>> class _UniversalSet(frozenset):
...     def __contains__(self, other):
...         return True

>>> 3 in _UniversalSet()
True

>>> %%timeit universe = frozenset()
>>> for num in range(100000):
...     _ = num in universe
1.47 ms ± 34 μs per loop

>>> %%timeit universe = frozenset()
>>> for num in range(100000):
...     _ = not universe or num in universe
1.47 ms ± 43.3 μs per loop

>>> %%timeit universe = _UniversalSet()
>>> for num in range(100000):
...     _ = num in universe
3.78 ms ± 40 μs per loop
```

Reply:
Remember, Python is not a fast language. 

This seems to show that the absolute difference for an individual invocation is about 2 nanoseconds. The example does not demonstrate a hotspot in a function of interest, and it's not clear whether the example is correct as CPython could easily implement a short circuit here making this an apples / oranges comparison. In the absence of evidence this very specific operation should be incrementally faster, isn't the readable and more explicit solution better?

Reply:
For this case I tend to respectfully disagree. In my opinion, creating a class just to solve a very simple problem is unnecessarily convoluted. The current method indeed needs performance, because trees can be large. Additionally tests without a potential CPython short circuit still show that class is significantly slower.

```python
>>> falses = [False] * 100000
>>> %%timeit universe = frozenset()
>>> for num in range(100000):
>>>     _ = not falses[num] or num in universe
1.77 ms ± 11.8 μs per loop
```